File: modules/whatz-it-live-overlay/ios/HybridLiveOverlayOutputFactory.swift
import Foundation
import NitroModules
import VisionCamera

final class HybridLiveOverlayOutputFactory: HybridLiveOverlayOutputFactorySpec {
  private var output: HybridLiveOverlayOutput?

  var isRecording: Bool {
    output?.isRecording ?? false
  }

  func createLiveOverlayOutput() throws -> any HybridCameraOutputSpec {
    let output = HybridLiveOverlayOutput()
    self.output = output
    return output
  }

  func startRecording(headshotPath: String?, wordmarkPath: String?) throws -> Promise<Void> {
    guard let output else {
      return .rejected(withError: makeError("The live-overlay camera output is unavailable."))
    }
    return try output.startRecording(headshotPath: headshotPath, wordmarkPath: wordmarkPath)
  }

  func appendOverlayEvent(event: LiveOverlayEvent) throws {
    guard let output else { return }
    try output.appendOverlayEvent(event: event)
  }

  func stopRecording() throws -> Promise<LiveOverlayRecordingResult> {
    guard let output else {
      return .rejected(withError: makeError("The live-overlay camera output is unavailable."))
    }
    return try output.stopRecording()
  }

  func cancelRecording() throws -> Promise<Void> {
    guard let output else { return .resolved() }
    return try output.cancelRecording()
  }

  private func makeError(_ message: String) -> Error {
    NSError(
      domain: "WhatzItLiveOverlay",
      code: 1,
      userInfo: [NSLocalizedDescriptionKey: message]
    )
  }
}

File: modules/whatz-it-live-overlay/android/src/main/java/com/margelo/nitro/whatzit/liveoverlay/HybridLiveOverlayOutputFactory.kt
package com.margelo.nitro.whatzit.liveoverlay

import com.margelo.nitro.camera.HybridCameraOutputSpec
import com.margelo.nitro.core.Promise
import com.whatzit.liveoverlay.HybridLiveOverlayOutput

class HybridLiveOverlayOutputFactory : HybridLiveOverlayOutputFactorySpec() {
  private var output: HybridLiveOverlayOutput? = null

  override val isRecording: Boolean
    get() = output?.isRecording ?: false

  override fun createLiveOverlayOutput(): HybridCameraOutputSpec {
    val nextOutput = HybridLiveOverlayOutput()
    output = nextOutput
    return nextOutput
  }

  override fun startRecording(headshotPath: String?, wordmarkPath: String?): Promise<Unit> =
    output?.startRecording(headshotPath, wordmarkPath)
      ?: Promise.async { throw IllegalStateException("The Android live-overlay camera output is unavailable.") }

  override fun appendOverlayEvent(event: LiveOverlayEvent) {
    output?.appendOverlayEvent(event)
  }

  override fun stopRecording(): Promise<LiveOverlayRecordingResult> =
    output?.stopRecording()
      ?: Promise.async { throw IllegalStateException("The Android live-overlay camera output is unavailable.") }

  override fun cancelRecording(): Promise<Unit> =
    output?.cancelRecording() ?: Promise.resolved()
}
